pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go

```diff
@@ -50,38 +50,112 @@ func ApplyClusterMerge(pctx networking.EnvoyFilter_PatchContext, efw *model.Envo
 			continue
 		}
 		if commonConditionMatch(pctx, cp) && clusterMatch(c, cp, hosts) {
+			// backup the Cluster in case mergeTransportSocketCluster modifies it
+			cpValueBkp := proto.Clone(cp.Value)
```

Comment:
ideally we would only clone when we are actually mutating things, to avoid high cost for users not using this functionality

Reply:
Patches aren't supposed to be terribly big, so this shouldn't be an issue. I suspect patching is not being called frequently too. How would you check if clone is needed? As I don't see simple solution and I think over-complicating already complicated code will probably not end well.

Reply:
Patching is called frequently and most of the CPU time tends to be on cloning

Reply:
In mergeTransportSocketCluster we are just overwriting TransportSocket or TransportSocketMatches, I think we can just shallowcopy it instead of deepcopy. This can be done in `protomarshal.ShallowCopy`, which should be much cheaper. I would also do it inside `mergeTransportSocketCluster`, so we don't to restore, I would make mergeTransportSocketCluster return the new patch and not modify the original cp.Value at all; modifying input params is always a little hard to follow and easy to make mistakes.

Reply:
Seems like an idea. Let's try.

Reply:
Seems like an idea. Let's try.